// redux-filter/lib/actions/creators.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.toggleFilter = toggleFilter;
exports.toggleOnly = toggleOnly;
exports.clearFilters = clearFilters;
exports.keywordSearch = keywordSearch;
exports.applySort = applySort;
exports.clearAllFilters = clearAllFilters;
exports.goToPage = goToPage;
exports.updateSubjects = updateSubjects;
exports.init = init;

var _constants = require('../constants.js');

function toggleFilter(attribute, value) {
    return {
        type: _constants.TOGGLE_FILTER,
        filter: { attribute: attribute, value: value }
    };
}

function toggleOnly(attribute, value) {
    return {
        type: _constants.TOGGLE_FILTER_ONLY,
        filter: { attribute: attribute, value: value }
    };
}

function clearFilters(attribute) {
    return {
        type: _constants.CLEAR_FILTERS,
        filter: { attribute: attribute }
    };
}

function keywordSearch(search) {
    return {
        type: _constants.KEYWORD_SEARCH,
        search: search
    };
}

function applySort(func) {
    return {
        type: _constants.APPLY_SORT,
        func: func
    };
}

function clearAllFilters() {
    return {
        type: _constants.CLEAR_ALL_FILTERS
    };
}

function goToPage(page) {
    return {
        type: _constants.GO_TO_PAGE,
        page: page
    };
}

function updateSubjects(subjects) {
    return {
        type: _constants.UPDATE_SUBJECTS,
        subjects: subjects
    };
}

function init() {
    return {
        type: _constants.INIT
    };
}
